### PR TITLE
Jonmv/reindexing performance test

### PR DESCRIPTION
**Pull request by jonmv**
@bjorncs @geirst @vekterli you may take a quick look if you want, but I still consider this a draft. Probably won't run for a long time still :D

### CODE REVIEW

**Comment by jonmv** [comment]
No longer a draft — PTAL :)

**Comment by jonmv** [comment]
Will of course squash.

File: lib/node_server.rb [hunk 1]
```diff
@@ -538,6 +538,22 @@ def wait_until_file_exists(filename, timeout)
...
   end
 
+  # Writes _type_ operations to generated document IDs _prefix[0, ..._count_], with other fields as given by _operation_ to _filename_.
+  # Example:
+  # write_document_operations(:update,
+  #                           { :condition => 'title="empty"', :fields => { :title => { :assign => 'unknown' } } },
+  #                           'id:ns:doctype::',
+  #                           1 << 20,
+  #                           '1M_ops.json')
+  def write_document_operations(type, operation, prefix, count, filename)
```
Comment (bjorncs) [inline lib/node_server.rb:548]: Nice 👍

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+    # Warmup and feed corpus
+    puts "Feeding initial data"
+    feed_data({ :file => initial_file })
+
+    # Benchmark pure reindexing
+    puts "Reindexing corpus"
+    sleep 2
+    now_seconds = Time.now.to_i
+    assert_hitcount("indexed_at_seconds:%3C#{now_seconds}&nocache", document_count)	# All documents should be indexed before now_seconds
```
Comment (bjorncs) [inline tests/performance/reindexing/reindexing_and_feeding.rb:109]: Consider making `%3C` and `%3E` a constant, or wrap raw string in url/percent-encoder

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+    deploy_app(@app)
+    wait_for_reindexing_start(current_reindexing_timestamp)
+  end
+
+  # Application and tenant names change based on the context this is run in.
+  def application_url
+    tenant = use_shared_configservers ? @tenant_name : "default"
+    application = use_shared_configservers ? @application_name : "default"
+    "https://#{vespa.nodeproxies.first[1].addr_configserver[0]}:#{19071}/application/v2/tenant/#{tenant}/application/#{application}/environment/prod/region/default/instance/default/"
```
Comment (bjorncs) [inline tests/performance/reindexing/reindexing_and_feeding.rb:250]: Set scheme to `http` to allow test to run without Vespa mTLS (underlying http(s) client will use correct protocol based on Vespa TLS settings).

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+  # Trigger reindexing of the whole corpus
+  def trigger_reindexing
+    # Read baseline reindexing status — very first reindexing is a no-op in the reindexer controller
+    response = http_request(URI(application_url + "reindexing"), {})
+    assert(response.code.to_i == 200, "Request should be successful")
+    previous_reindexing_timestamp = get_json(response)["status"]["readyMillis"]
+
+    # Allow the reindexing maintainer some time to run, and mark the first no-op reindexing as done
+    sleep 60
```
Comment (bjorncs) [inline tests/performance/reindexing/reindexing_and_feeding.rb:230]: Can this be replaced by periodic polling of some status API?

**Review by bjorncs — APPROVED** [review]

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 4 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:230]: Oops, leftover from old code.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 3 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:250]: OK.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 2 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:109]: Will consider.

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 5]
```diff
@@ -0,0 +1,255 @@
+
+# coding: utf-8
+# Copyright 2020 Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+require 'performance_test'
+require 'app_generator/search_app'
+
+class ReindexingAndFeeding < PerformanceTest
```
Comment (geirst) [inline tests/performance/reindexing/reindexing_and_feeding.rb:8]: Consider 'Test' suffix on class name.

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+  def initialize(*args)
+    super(*args)
+  end
+
+  def timeout_seconds
+    900
+  end
+
+  def setup
+	  super
```
Comment (geirst) [inline tests/performance/reindexing/reindexing_and_feeding.rb:19]: Indentation is off. Should be 2 spaces. Applies for lines further down as well.

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+				       updates_file)
+
+    # Warmup and feed corpus
+    puts "Feeding initial data"
+    feed_data({ :file => initial_file })
+    assert_hitcount("sddocname:doc", document_count) # All documents should be fed and visible
+
+    # Benchmark pure reindexing
+    puts "Reindexing corpus"
+    sleep 2
```
Comment (geirst) [inline tests/performance/reindexing/reindexing_and_feeding.rb:108]: Is this really needed?

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    start
+
+    # First time is a dummy.
+    trigger_reindexing
+    wait_for_reindexing
+
+    qrserver = @vespa.container["combinedcontainer/0"]
+    document_count = 1 << 20
+
+    initial_file = dirs.tmpdir + "initial.json"
```
Comment (geirst) [inline tests/performance/reindexing/reindexing_and_feeding.rb:77]: I suggest moving all the data generation to a separate function and set the files as private variables with '@..'. This makes it easier to read to overall flow of the test.

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+				       'id:test:doc::',
+				       document_count,
+				       updates_file)
+
+    # Warmup and feed corpus
+    puts "Feeding initial data"
+    feed_data({ :file => initial_file })
+    assert_hitcount("sddocname:doc", document_count) # All documents should be fed and visible
+
+    # Benchmark pure reindexing
```
Comment (geirst) [inline tests/performance/reindexing/reindexing_and_feeding.rb:106]: I suggest creating a separate function for each of the test cases. Makes it easier to read the test in my opinion.

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+  def timeout_seconds
+    900
+  end
+
+  def setup
+	  super
+	  set_description("Measure throughput of reindexing, and its impact on external updates and puts")
+	  set_owner("jvenstad")
+	  @test_config = {
```
Comment (geirst) [inline tests/performance/reindexing/reindexing_and_feeding.rb:22]: Are the values in this map updated for this test?

File: tests/performance/reindexing/reindexing_and_feeding.rb [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    end
+    assert(Time.now - start_time < 60, "Services should converge on new generation within the minute")
+    assert(generation == get_json(http_request(URI(application_url + "serviceconverge"), {}))["wantedGeneration"],
+           "Should converge on generation #{generation}")
+    puts "Services converged on new config generation after #{Time.now - start_time} seconds"
+  end
+
+  # Trigger reindexing of the whole corpus
+  def trigger_reindexing
+    # Read baseline reindexing status — very first reindexing is a no-op in the reindexer controller
```
Comment (geirst) [inline tests/performance/reindexing/reindexing_and_feeding.rb:225]: Consider if some of the code here can be shared with similar code in schemachanges_need_refeed_reconfig.rb.

**Review by geirst — COMMENTED** [review]
Overall looks good, some comments to look at.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 5 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:8]: Will change.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 6 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:19]: Argh. Will fix.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 7 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:108]: Not sure. Test failed locally without it, but that _could_ be a random occurrence. Can try removing it when test is stable.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 8 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:77]: Can do. Yes, it's a bit verbose.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 9 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:106]: Can do. I generally prefer to keep more things in (mental) scope, but I'm starting to see that I'm the outlier here 😖

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 10 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:22]: Oops 😨 Forgot all about the graphs.

File: tests/performance/reindexing/reindexing_and_feeding.rb (hunk 11 above)
Comment (jonmv) [reply tests/performance/reindexing/reindexing_and_feeding.rb:225]: Already did that. Not much more to share. That test is a bit different, so I couldn't easily find such opportunities.

**Pull request merged**